### ISSUE TITLE
Distinguish user value changes from programatic ones

### ISSUE THREAD

**Issue by skogsmaskin** [sghall/react-compound-slider#42]
### Problem or feature description:

First off, thanks for this nice module, it's really awesome. 

I'm having an issue where it's hard to track wether the user changed the value, or it changed from the outside (via props). Think of a use case where you want to use the slider to act as a video seek bar. The handle should track the video current time (from outside prop), but the user should be able to modify it. While the user is modifying it,  the handle should not track the outside prop and respect the user's input values (onChange, onUpdate).

I was trying to use `onSlideStart/End` to set a state `isSeeking` and not track the outside prop while this state is true, **but it's not consistent with changing the value by clicking on the rail**.

One idea to solve this could be to also include clicking on the rail (and thus sliding) in the `onSlideStart` and `onSlideEnd` callbacks too, but one would loose exclusivity for drag events.

Another approach would be to introduce some new callbacks that includes all user events. Something like `onInputStart(event, value)` `onInputEnd(event, value)`.

What do you think? I could have a stab at a PR, but the API should be decided first.

### Showcase

Try clicking on the rail to reposition handle vs dragging the handle.

https://codesandbox.io/s/y3mo40wqv1

**Comment by sghall** [comment]
Hey there.  Yeah, interesting use case.  Never tried anything like that before.  You also have the "onUpdate"  which will give you everything and you need to pick out the cases you care about.

```
  public onUpdate = values => {
    this.setState(({ isSeeking }) => {
      if (isSeeking) {
        return null;
      }

      this.props.onSeekTo(values[0]);
      return { values };
    });
  };
```

I think this does what you want it to do. Right? 
https://codesandbox.io/s/qz315pn119

**Comment by skogsmaskin** [comment]
Yes, but when it is an actual video (and not a Javascript Interval) you will get into trouble because ``onSeek`` is actually ``video.currentTime=value[0]``.

**Comment by skogsmaskin** [comment]
Here is an example with an actual video demonstrating the issue:

https://codesandbox.io/s/xl89j03vpz

**Comment by skogsmaskin** [comment]
By adding my own onClick slider handler I got it working:

https://codesandbox.io/s/q78wqv12y6

``
  static getDerivedStateFromProps(props, state) {
    if (!state.isSeeking) {
      return {
        values: [props.currentTime || 0]
      };
    }
    return null;
  }

  public componentDidUpdate(prevProps, prevState) {
    if (prevState.isSeeking && !this.state.isSeeking) {
      this.props.onSeekTo(prevState.values[0]);
    }
  }

  public onChange = (values: number[]) => {
    this.setState({ values });
  };

  public onUpdate = (values: number[]) => {
    this.setState({ values });
  };

  public handleSlideStart = () => {
    this.setState({ isSeeking: true });
  };

  public handleSlideEnd = () => {
    this.setState({ isSeeking: false });
  };

  public handleSliderClick = () => {
    this.setState({ isSeeking: true });
    setTimeout(() => {
      this.props.onSeekTo(this.state.values[0]);
    }, 100);
  };
``

**Comment by skogsmaskin** [comment]
There is another issue, seems like everything I put over or under the slider will only take `mouseup`. Maybe the slider is event.preventDefaulting it?

Anyway, if adding a click handler to the module, this kind of functionality will be easy.

**Comment by sghall** [comment]
Yeah,  cool.  If you want to put together a PR that's great.  I'm traveling right now, but go ahead and submit it. Should be able to take a look in a few days. Cheers.

**Comment by skogsmaskin** [comment]
Cool, I've submitted a PR. My problem can now be solved with this code:

```
  static getDerivedStateFromProps(props, state) {
    if (!state.isSeeking) {
      return {
        values: [props.currentTime || 0]
      };
    }
    return null;
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.isSeeking && !this.state.isSeeking) {
      this.props.onSeekTo(prevState.values[0]);
    }
  }

  handleSliderChange = (values: number[]) => {
    this.setState({ values });
  };

  handleSliderUpdate = (values: number[]) => {
    this.setState({ values });
  };

  handleSlideStart = () => {
    this.setState({ isSeeking: true });
  };

  handleSlideEnd = () => {
    this.setState({ isSeeking: false });
  };

  handlePressRail = () => {
    this.setState({ isSeeking: true }, () => {
      this.setState({ isSeeking: false });
    });
  };
```

**Comment by atlanteh** [comment]
Actually I think that according to React standard, onChange should only be called when user changed the input. Just like any <Input /> "onChange" event is called only when user changes the input and not when the prop was changed. we can leave onUpdate event to be called upon any change

**Comment by skogsmaskin** [comment]
@sghall - yeah, I was also able to do this if everything was in one component. However, I was not able to do it if the UI was supposed to be in one component and the video in another, and that's what I was trying to solve.

**Comment by sghall** [comment]
I see.  Well, for the slider part, it's just moving it into a separate file like this...
https://codesandbox.io/s/1z51zn6q23

**Comment by skogsmaskin** [comment]
@sghall - cool, that seems to work fine! I suppose adding the state ``videoCursor`` was the trick (and not directly tie it to ``video.currentTime``)?

**Comment by sghall** [comment]
Yeah, seems to me, you need to track the time in two different places.  This is what I hacked together.  It seems to work well, but be careful I really don't know much working with videos and had to hit MDN a few times.

There's also one really key thing which I am exploiting: the onUpdate will ALWAYS be called before the onChange.  Here in the source:

https://github.com/sghall/react-compound-slider/blob/7b7d01a45326f406abd3d5d1089d8960acb01d7e/src/Slider/Slider.js#L388

If you look closely the sandbox this is a key assumption.  When you let go of the slider it calls onUpdate and then onChange (setting videoCursor back to null etc).  Just something to take note off.  So I can understand the confusion in getting something like this to work.

**Comment by sghall** [comment]
@thunderkid correctly pointed out that the videoCursor is not needed in https://github.com/sghall/react-compound-slider/issues/47.  So this is most up-to-date version with some fixes to the styles etc:
https://codesandbox.io/s/64vz7m7v6z

**Comment by thunderkid** [comment]
Minor tweak - this version:
[https://codesandbox.io/s/4303ylv5n9](https://codesandbox.io/s/4303ylv5n9)
just adds `onEnded` to @sghall's to keep the play/pause buttons in the right state when the video ends.

**Comment by sghall** [comment]
Closing this one.  I think we got this.  Open it back up if there's still an issue here.
